### PR TITLE
[5.0] Use custom page name for Paginator

### PR DESCRIPTION
**Pull request by sahibalejandro**
Resolving issue #8000

### CODE REVIEW

File: src/Illuminate/Database/Eloquent/Relations/BelongsToMany.php [hunk 1]
```diff
@@ -173,13 +173,14 @@ public function get($columns = array('*'))
 	 *
 	 * @param  int    $perPage
 	 * @param  array  $columns
+	 * @param  string $pageName
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToMany.php]: cs

File: src/Illuminate/Database/Eloquent/Relations/HasManyThrough.php [hunk 2]
```diff
@@ -304,13 +304,14 @@ protected function getSelectColumns(array $columns = ['*'])
 	 *
 	 * @param  int    $perPage
 	 * @param  array  $columns
+	 * @param  string $pageName
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/HasManyThrough.php]: cs

File: src/Illuminate/Database/Query/Builder.php [hunk 3]
```diff
@@ -1380,13 +1380,14 @@ protected function runSelect()
 	/**
 	 * Paginate the given query into a simple paginator.
 	 *
-	 * @param  int  $perPage
-	 * @param  array  $columns
+	 * @param  int $perPage
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Query/Builder.php]: cs

File: src/Illuminate/Database/Query/Builder.php [hunk 4]
```diff
@@ -1380,13 +1380,14 @@ protected function runSelect()
...
+	 * @param  array $columns
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Query/Builder.php]: cs

File: src/Illuminate/Database/Query/Builder.php [hunk 5]
```diff
@@ -1380,13 +1380,14 @@ protected function runSelect()
...
+	 * @param  string $pageName
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Query/Builder.php]: cs

File: src/Illuminate/Database/Query/Builder.php [hunk 6]
```diff
@@ -1402,18 +1404,20 @@ public function paginate($perPage = 15, $columns = ['*'])
 	 *
 	 * This is more efficient on larger data-sets, etc.
 	 *
-	 * @param  int  $perPage
-	 * @param  array  $columns
+	 * @param  int $perPage
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Query/Builder.php]: cs

File: src/Illuminate/Database/Query/Builder.php [hunk 7]
```diff
@@ -1402,18 +1404,20 @@ public function paginate($perPage = 15, $columns = ['*'])
...
+	 * @param  array $columns
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Query/Builder.php]: you get the point...

Comment (sahibalejandro) [inline src/Illuminate/Database/Query/Builder.php]: Ham... sorry, what is "cs" ?
This PR is good or bad?

Comment (GrahamCampbell) [inline src/Illuminate/Database/Query/Builder.php]: Coding style is wrong. Look at every other docblock in the framework.

**Comment by sahibalejandro** [comment]
I think it's ok now.

File: src/Illuminate/Database/Eloquent/Builder.php [hunk 8]
```diff
@@ -239,34 +239,37 @@ public function lists($column, $key = null)
 	/**
 	 * Paginate the given query.
 	 *
-	 * @param  int  $perPage
-	 * @param  array  $columns
+	 * @param  int     $perPage
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Builder.php]: don't align please - just use 2 spaces

**Comment by GrahamCampbell** [comment]
Please squash to one commit when you're done.

**Comment by sahibalejandro** [comment]
Oh sorry, I did not read about the squash commit. :(

**Comment by taylorotwell** [comment]
What is the use case for using a custom name for this?

**Comment by sahibalejandro** [comment]
It's necessary when you have to paginate two or more results in the same page, for example, on a site I'm building, I need to show two list of products, "Featured Products" and "Related Products", each list has their own pagination links, and of course they need a different "page" variable to avoid conflicts.

When I was looking for a solution, I found many other developers need this advantage, so I decide to make the pull request.

The issue is here #8000 but was closed without solution, and here is "the possible bug" https://github.com/illuminate/pagination/blob/master/PaginationServiceProvider.php#L21

**Comment by taylorotwell** [comment]
Can you rebase and fix merge conflicts.

**Comment by sahibalejandro** [comment]
I think it's ready.

**Comment by GrahamCampbell** [comment]
That's not correct. You've rebased against the wrong branch. We can't merge this.

**Comment by GrahamCampbell** [comment]
You need to actually REBASE and force push. There MUST be only one commit in here.

**Comment by sahibalejandro** [comment]
OMG... I'm confused with "rebasing" things...

**Comment by GrahamCampbell** [comment]
5.0, then force push. There must be exactly one commit here, not loads of them.

**Comment by sahibalejandro** [comment]
It's now correct?

**Comment by GrahamCampbell** [comment]
Nearly, just squash and force push and this should be good. :)

File: src/Illuminate/Pagination/AbstractPaginator.php [hunk 9]
```diff
@@ -305,14 +305,15 @@ public static function currentPathResolver(Closure $resolver)
 	/**
 	 * Resolve the current page or return the default value.
 	 *
+	 * @param  string  $pageName
 	 * @param  int  $default
 	 * @return int
 	 */
-	public static function resolveCurrentPage($default = 1)
+	public static function resolveCurrentPage($pageName = 'page', $default = 1)
```
Comment (GrahamCampbell) [inline src/Illuminate/Pagination/AbstractPaginator.php]: this is a breaking change to the public api

**Comment by GrahamCampbell** [comment]
@sahibalejandro No need for you to do anything, I'll get this sorted for you. :)

**Comment by GrahamCampbell** [comment]
Merged into 5.1. Thank you @sahibalejandro.

File: src/Illuminate/Pagination/AbstractPaginator.php (hunk 9 above)
Comment (sahibalejandro) [inline src/Illuminate/Pagination/AbstractPaginator.php]: I noticed this, I searched all calls to `resolveCurrentPage()` on the code and none is passing that parameter yet.

**Comment by sahibalejandro** [comment]
Nice, thank you @GrahamCampbell, and sorry for the little mess.

**Comment by GlashutteOriginal** [comment]
I don't know, why I changed all the codes you have commit , but two paginations in one page didn't work, and what should i do to make it works?

File: src/Illuminate/Database/Query/Builder.php (hunk 7 above)
Comment (CaddyDz) [reply src/Illuminate/Database/Query/Builder.php]: @GrahamCampbell that is the most arrogant condescending review I have ever seen!  
Grow up please

**Pull request closed without merging**